design/local-node-agent.md

```diff
@@ -0,0 +1,123 @@
+For example, a Flexvolume plugin does not execute in a context that has cluster credentials, so it cannot communicate with the Kubernetes API to perform such operations as creating Custom Resource Definitions (CRDs).
+This document will describe how a Rook local node agent would work with the Flexvolume architecture to provide storage in a Kubernetes cluster, in addition to other responsibilities that must be performed on specific nodes in the cluster.
+
+## **Detailed Design**
+
+### **Responsibilities**
+The Rook agent will have multiple responsibilities beyond just performing storage operations for the Rook Flexvolume driver.
+One can think of the agent as a mini operator that functions at the node level.
+The initial proposed responsibilities of the agent are:
+
+1. Deploy the Rook Flexvolume driver to the `volume-plugin-dir` directory on every node
+1. Perform storage operations on behalf of the Flexvolume driver, such as attaching, detaching, mounting and unmounting Rook cluster storage
+1. Cluster clean up operations, such as forcefully unmapping RBD devices when the Rook cluster is being deleted while there are still pods consuming those volumes
+1. Proxy I/O traffic from kernel modules to user space (e.g. Network Block Device (NBD) kernel module to librbd in userspace)
+
+### **Deployment**
+The Rook operator will deploy the Rook agent to all nodes in the cluster via a Daemonset in the same namespace in which the operator is running.
+It is a permanent (long running) daemon that has a lifecycle tied to the node that it is scheduled on.
+The agent deployment will happen when the operator is first created, in the same flow where the operator is declaring CRDs for clusters, pools, etc.
+This means that the Rook agents are not associated with a specific Rook cluster and that they will be able to handle operations for any Rook cluster instance.
+The Cluster CRD will be updated to allow selectors to control the set of nodes that the agent is scheduled on, but in most cases it is desirable for it to be running on all nodes.
+
+Each agent pod will be running the same `rook/rook` container image in use today, but with a new `agent` command (similar to the existing `mon` or `osd` commands).
+When the agent starts up, it will immediately copy the Rook Flexvolume driver from its container image to the `volume-plugin-dir` on its host node.
+In Kubernetes 1.8+, the [dynamic Flexvolume plugin discovery](https://github.com/kubernetes/community/pull/833) will find and initialize our driver, but in older versions of Kubernetes a manual restart of the Kubelet will be required.
+
+After the driver has been deployed, the Rook agent will create a Unix domain socket that will serve as the communications channel between the agent and the Flexvolume driver.
+The agent will wait for the driver to complete a handshake over the socket before moving on and being ready to accept Flexvolume storage requests.
```

Comment:
it makes sense to start the handshake from `Init()`, i've clarified that in the flow.